File: integration_tests/rust/src/circular_dependency.rs
use {
    crate::{
        assertions::assert_round_trip,
        types::circular_dependency::{
            dependency::types::{StructFromBelowIn, StructFromBelowOut},
            types::{StructFromAboveIn, StructFromAboveOut},
        },
    },
    std::io,
};

pub fn run() -> io::Result<()> {
    assert_round_trip::<StructFromAboveOut, StructFromAboveIn>(&StructFromAboveOut {
        field: "field".to_owned(),
        size: "size".to_owned(),
        elements: "elements".to_owned(),
        fallback: "fallback".to_owned(),
    })?;

    println!();

    assert_round_trip::<StructFromBelowOut, StructFromBelowIn>(&StructFromBelowOut {
        x: StructFromAboveOut {
            field: "field".to_owned(),
            size: "size".to_owned(),
            elements: "elements".to_owned(),
            fallback: "fallback".to_owned(),
        },
    })?;

    Ok(())
}

File: integration_tests/typescript/src/circular-dependency.ts
import { CircularDependency } from '../generated/types';
import { assertRoundTrip } from './assertions';

export default function run(): void {
  assertRoundTrip(
    CircularDependency.Types.StructFromAbove.size,
    CircularDependency.Types.StructFromAbove.serialize,
    CircularDependency.Types.StructFromAbove.deserialize,
    {},
  );

  // eslint-disable-next-line no-console
  console.log();

  assertRoundTrip(
    CircularDependency.Dependency.Types.StructFromBelow.size,
    CircularDependency.Dependency.Types.StructFromBelow.serialize,
    CircularDependency.Dependency.Types.StructFromBelow.deserialize,
    { x: {} },
  );
}
